File: ClipmanMac/Sources/ClipmanCore/SyncConflictResolver.swift
import Foundation
import CryptoKit

public enum SyncConflictResolver {
    public static func resolveDatabaseConflicts(databaseURL: URL, password: String) throws -> Bool {
        let conflicts = conflictSiblings(for: databaseURL)
        guard !conflicts.isEmpty else { return false }

        var merged = ClipDatabase()
        if FileManager.default.fileExists(atPath: databaseURL.path) {
            merge(into: &merged, source: try ClipDatabaseFile.load(databaseURL, password: password))
        }
        for conflict in conflicts {
            merge(into: &merged, source: try ClipDatabaseFile.load(conflict, password: password))
        }

        normalize(&merged)
        try ClipDatabaseFile.saveAtomic(databaseURL, database: merged, password: password)
        for conflict in conflicts {
            try? FileManager.default.removeItem(at: conflict)
        }
        return true
    }

    public static func merge(into target: inout ClipDatabase, source: ClipDatabase) {
        _ = mergeDeletedEntries(into: &target, source: source)
        applyDeletedEntries(&target)
        for incoming in source.Entries where !incoming.Text.isEmpty {
            if isDeleted(incoming, in: target) { continue }
            if let idIndex = target.Entries.firstIndex(where: { !$0.Id.isEmpty && $0.Id.caseInsensitiveCompare(incoming.Id) == .orderedSame }) {
                mergeEntry(existing: &target.Entries[idIndex], incoming: incoming)
                continue
            }
            if let textIndex = target.Entries.firstIndex(where: { $0.Text == incoming.Text }) {
                mergeEntry(existing: &target.Entries[textIndex], incoming: incoming)
                continue
            }
            target.Entries.append(incoming)
        }
        applyDeletedEntries(&target)
    }

    public static func normalize(_ database: inout ClipDatabase) {
        database.Version = max(1, database.Version)
        database.UpdatedUnixMs = TimeUtil.nowUnixMs()
        normalizeDeletedEntries(&database)
        applyDeletedEntries(&database)
        let orderedIDs = database.Entries
            .sorted {
                let leftOrder = $0.ManualOrder <= 0 ? Int64.max : $0.ManualOrder
                let rightOrder = $1.ManualOrder <= 0 ? Int64.max : $1.ManualOrder
                if leftOrder == rightOrder { return $0.CreatedUnixMs < $1.CreatedUnixMs }
                return leftOrder < rightOrder
            }
            .map(\.Id)
        for (offset, id) in orderedIDs.enumerated() {
            guard let index = database.Entries.firstIndex(where: { $0.Id == id }) else { continue }
            if database.Entries[index].Id.isEmpty {
                database.Entries[index].Id = UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
            }
            if database.Entries[index].CreatedUnixMs == 0 {
                database.Entries[index].CreatedUnixMs = TimeUtil.nowUnixMs()
            }
            if database.Entries[index].LastUsedUnixMs == 0 {
                database.Entries[index].LastUsedUnixMs = database.Entries[index].CreatedUnixMs
            }
            database.Entries[index].ManualOrder = Int64(offset + 1)
        }
    }

    public static func isDeleted(_ id: String, in database: ClipDatabase) -> Bool {
        guard !id.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
        return database.DeletedEntries.contains { $0.Id == id }
    }

    public static func isDeleted(_ entry: ClipEntry, in database: ClipDatabase) -> Bool {
        isDeleted(entry, deletedEntries: database.DeletedEntries)
    }

    private static func isDeleted(_ entry: ClipEntry, deletedEntries: [DeletedClipEntry]) -> Bool {
        if deletedEntries.contains(where: { $0.Id == entry.Id }) { return true }
        guard !entry.Text.isEmpty else { return false }
        let hash = textHash(entry.Text)
        let entryChangedUnixMs = max(entry.CreatedUnixMs, entry.LastUsedUnixMs)
        return deletedEntries.contains {
            !$0.TextHash.isEmpty
                && $0.TextHash == hash
                && ($0.DeletedUnixMs <= 0 || entryChangedUnixMs <= $0.DeletedUnixMs)
        }
    }

    @discardableResult
    public static func mergeDeletedEntries(into target: inout ClipDatabase, source: ClipDatabase) -> Bool {
        normalizeDeletedEntries(&target)
        var normalizedSource = source
        normalizeDeletedEntries(&normalizedSource)
        guard !normalizedSource.DeletedEntries.isEmpty else { return false }

        var changed = false
        for deleted in normalizedSource.DeletedEntries where !deleted.Id.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            if let index = target.DeletedEntries.firstIndex(where: { $0.Id == deleted.Id }) {
                if deleted.DeletedUnixMs > target.DeletedEntries[index].DeletedUnixMs {
                    target.DeletedEntries[index] = deleted
                    changed = true
                } else if target.DeletedEntries[index].TextHash.isEmpty && !deleted.TextHash.isEmpty {
                    target.DeletedEntries[index].TextHash = deleted.TextHash
                    changed = true
                }
            } else {
                target.DeletedEntries.append(deleted)
                changed = true
            }
        }
        normalizeDeletedEntries(&target)
        return changed
    }

    public static func addDeletedEntry(id: String, text: String, machineName: String, to database: inout ClipDatabase) {
        let trimmed = id.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        let marker = DeletedClipEntry(Id: trimmed, TextHash: textHash(text), DeletedUnixMs: TimeUtil.nowUnixMs(), SourceMachine: machineName)
        if let index = database.DeletedEntries.firstIndex(where: { $0.Id == trimmed }) {
            database.DeletedEntries[index] = marker
        } else {
            database.DeletedEntries.append(marker)
        }
        normalizeDeletedEntries(&database)
        applyDeletedEntries(&database)
    }

    public static func applyDeletedEntries(_ database: inout ClipDatabase) {
        guard !database.DeletedEntries.isEmpty else { return }
        let deletedEntries = database.DeletedEntries
        database.Entries.removeAll { isDeleted($0, deletedEntries: deletedEntries) }
    }

    private static func normalizeDeletedEntries(_ database: inout ClipDatabase) {
        let cutoff = TimeUtil.nowUnixMs() - Int64(90 * 24 * 60 * 60 * 1000)
        var byID: [String: DeletedClipEntry] = [:]
        for marker in database.DeletedEntries {
            let id = marker.Id.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !id.isEmpty else { continue }
            guard marker.DeletedUnixMs == 0 || marker.DeletedUnixMs >= cutoff else { continue }
            var normalized = marker
            normalized.Id = id
            if normalized.TextHash.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                normalized.TextHash = ""
            }
            if normalized.DeletedUnixMs == 0 {
                normalized.DeletedUnixMs = TimeUtil.nowUnixMs()
            }
            if let existing = byID[id], existing.DeletedUnixMs >= normalized.DeletedUnixMs {
                continue
            }
            byID[id] = normalized
        }
        database.DeletedEntries = byID.values.sorted {
            if $0.DeletedUnixMs == $1.DeletedUnixMs { return $0.Id < $1.Id }
            return $0.DeletedUnixMs > $1.DeletedUnixMs
        }
    }

    public static func textHash(_ text: String) -> String {
        guard let data = text.data(using: .utf8), !data.isEmpty else { return "" }
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    public static func conflictSiblings(for canonicalURL: URL) -> [URL] {
        let directory = canonicalURL.deletingLastPathComponent()
        let baseName = canonicalURL.deletingPathExtension().lastPathComponent
        let fileExtension = canonicalURL.pathExtension
        guard let children = try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil) else {
            return []
        }
        return children.filter { url in
            url != canonicalURL &&
                url.pathExtension.caseInsensitiveCompare(fileExtension) == .orderedSame &&
                isConflictName(url.deletingPathExtension().lastPathComponent, baseName: baseName)
        }
    }

    private static func mergeEntry(existing: inout ClipEntry, incoming: ClipEntry) {
        let incomingLastUsed = incoming.LastUsedUnixMs
        let incomingWins = incomingLastUsed >= existing.LastUsedUnixMs
        let incomingCreatedWins = incoming.CreatedUnixMs > existing.CreatedUnixMs
        if incomingLastUsed > existing.LastUsedUnixMs {
            existing.LastUsedUnixMs = incomingLastUsed
        }
        if incoming.CreatedUnixMs > 0
            && (existing.CreatedUnixMs == 0
                || incomingCreatedWins
                || (!incomingWins && incoming.CreatedUnixMs < existing.CreatedUnixMs)) {
            existing.CreatedUnixMs = incoming.CreatedUnixMs
        }
        if !incoming.Name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && incomingWins {
            existing.Name = incoming.Name.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if !incoming.Group.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && incomingWins {
            existing.Group = incoming.Group.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if !incoming.SourceMachine.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && (incomingWins || incomingCreatedWins) {
            existing.SourceMachine = incoming.SourceMachine.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        existing.Pinned = existing.Pinned || incoming.Pinned
        if existing.ManualOrder <= 0 || (incoming.ManualOrder > 0 && incoming.ManualOrder < existing.ManualOrder) {
            existing.ManualOrder = incoming.ManualOrder
        }
    }

    private static func isConflictName(_ name: String, baseName: String) -> Bool {
        guard name.range(of: baseName, options: [.caseInsensitive, .anchored]) != nil else { return false }
        let suffix = name.dropFirst(baseName.count).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !suffix.isEmpty else { return false }
        let lower = suffix.lowercased()
        if lower.contains("conflicted copy") { return true }
        if lower.contains("[conflict]") { return true }
        if lower.contains(" conflict") { return true }
        if lower.hasPrefix("_conf(") { return true }
        if lower.hasPrefix(" _conf(") { return true }
        return looksLikeOneDriveComputerSuffix(String(suffix))
    }

    private static func looksLikeOneDriveComputerSuffix(_ suffix: String) -> Bool {
        var value = suffix.trimmingCharacters(in: .whitespacesAndNewlines)
        guard value.count >= 3 && value.count <= 80 else { return false }
        guard value.hasPrefix("-") || value.hasPrefix("(") || value.hasPrefix(" ") else { return false }
        value = value.trimmingCharacters(in: CharacterSet(charactersIn: " -()"))
        guard value.count >= 2 && value.count <= 64 else { return false }
        return value.allSatisfy { $0.isLetter || $0.isNumber || $0 == "-" || $0 == "_" }
    }
}

File: src/SyncConflictResolver.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

namespace Clipman
{
    internal static class SyncConflictResolver
    {
        public static void ResolveSettingsConflicts(string settingsPath)
        {
            try
            {
                var conflicts = FindConflictSiblings(settingsPath).ToList();
                if (conflicts.Count == 0) return;

                var allFiles = new List<string>();
                if (File.Exists(settingsPath)) allFiles.Add(settingsPath);
                allFiles.AddRange(conflicts);
                var newest = allFiles
                    .Select(path => new FileInfo(path))
                    .Where(info => info.Exists)
                    .OrderByDescending(info => info.LastWriteTimeUtc)
                    .FirstOrDefault();
                if (newest == null) return;

                Directory.CreateDirectory(Path.GetDirectoryName(settingsPath));
                if (!string.Equals(newest.FullName, settingsPath, StringComparison.OrdinalIgnoreCase))
                {
                    File.Copy(newest.FullName, settingsPath, true);
                }

                DeleteFiles(conflicts);
            }
            catch
            {
            }
        }

        public static void ResolveDatabaseConflicts(string databasePath)
        {
            ResolveDatabaseConflicts(databasePath, string.Empty);
        }

        public static void ResolveDatabaseConflicts(string databasePath, string password)
        {
            try
            {
                var conflicts = FindConflictSiblings(databasePath).ToList();
                if (conflicts.Count == 0) return;

                var merged = new ClipDatabase();
                if (File.Exists(databasePath))
                {
                    MergeInto(merged, ClipDatabaseFile.Load(databasePath, password));
                }

                foreach (var conflict in conflicts)
                {
                    MergeInto(merged, ClipDatabaseFile.Load(conflict, password));
                }

                NormalizeMergedDatabase(merged);
                ClipDatabaseFile.SaveAtomic(databasePath, merged, password);
                DeleteFiles(conflicts);
            }
            catch
            {
            }
        }

        public static IEnumerable<string> FindConflictSiblings(string canonicalPath)
        {
            if (string.IsNullOrWhiteSpace(canonicalPath)) yield break;
            var directory = Path.GetDirectoryName(canonicalPath);
            var fileName = Path.GetFileName(canonicalPath);
            if (string.IsNullOrWhiteSpace(directory) || string.IsNullOrWhiteSpace(fileName) || !Directory.Exists(directory))
            {
                yield break;
            }

            var baseName = Path.GetFileNameWithoutExtension(fileName);
            var extension = Path.GetExtension(fileName);
            foreach (var file in Directory.GetFiles(directory, "*" + extension))
            {
                if (string.Equals(file, canonicalPath, StringComparison.OrdinalIgnoreCase)) continue;
                if (IsConflictNameFor(file, baseName, extension))
                {
                    yield return file;
                }
            }
        }

        private static bool IsConflictNameFor(string path, string baseName, string extension)
        {
            if (!string.Equals(Path.GetExtension(path), extension, StringComparison.OrdinalIgnoreCase)) return false;
            var name = Path.GetFileNameWithoutExtension(path);
            if (string.IsNullOrWhiteSpace(name) || !name.StartsWith(baseName, StringComparison.OrdinalIgnoreCase)) return false;

            var suffix = name.Substring(baseName.Length).Trim();
            if (suffix.Length == 0) return false;
            var lower = suffix.ToLowerInvariant();

            if (lower.Contains("conflicted copy")) return true;
            if (lower.Contains("[conflict]")) return true;
            if (lower.Contains(" conflict")) return true;
            if (lower.StartsWith("_conf(")) return true;
            if (lower.StartsWith(" _conf(")) return true;

            return LooksLikeOneDriveComputerSuffix(suffix);
        }

        private static bool LooksLikeOneDriveComputerSuffix(string suffix)
        {
            var value = suffix.Trim();
            if (value.Length < 3 || value.Length > 80) return false;
            if (!(value.StartsWith("-", StringComparison.Ordinal) ||
                  value.StartsWith("(", StringComparison.Ordinal) ||
                  value.StartsWith(" ", StringComparison.Ordinal)))
            {
                return false;
            }

            value = value.Trim(' ', '-', '(', ')');
            if (value.Length < 2 || value.Length > 64) return false;
            return value.All(ch => char.IsLetterOrDigit(ch) || ch == '-' || ch == '_');
        }

        public static void MergeInto(ClipDatabase target, ClipDatabase source)
        {
            if (target == null || source == null || source.Entries == null) return;
            if (target.Entries == null) target.Entries = new List<ClipEntry>();
            if (target.DeletedEntries == null) target.DeletedEntries = new List<DeletedClipEntry>();
            MergeDeletedEntries(target, source);
            ApplyDeletedEntries(target);
            foreach (var entry in source.Entries.Where(e => e != null && !string.IsNullOrEmpty(e.Text)))
            {
                if (IsDeleted(target, entry)) continue;
                var existing = target.Entries.FirstOrDefault(e =>
                    !string.IsNullOrWhiteSpace(entry.Id) &&
                    string.Equals(e.Id, entry.Id, StringComparison.OrdinalIgnoreCase));
                if (existing == null)
                {
                    existing = target.Entries.FirstOrDefault(e => string.Equals(e.Text, entry.Text, StringComparison.Ordinal));
                }

                if (existing == null)
                {
                    target.Entries.Add(Clone(entry));
                    continue;
                }

                MergeEntry(existing, entry);
            }
            ApplyDeletedEntries(target);
        }

        private static void MergeEntry(ClipEntry existing, ClipEntry incoming)
        {
            var incomingWins = incoming.LastUsedUnixMs >= existing.LastUsedUnixMs;
            var incomingCreatedWins = incoming.CreatedUnixMs > existing.CreatedUnixMs;
            if (incoming.LastUsedUnixMs > existing.LastUsedUnixMs) existing.LastUsedUnixMs = incoming.LastUsedUnixMs;
            if (incoming.CreatedUnixMs > 0 &&
                (existing.CreatedUnixMs == 0 ||
                 incomingCreatedWins ||
                 (!incomingWins && incoming.CreatedUnixMs < existing.CreatedUnixMs)))
            {
                existing.CreatedUnixMs = incoming.CreatedUnixMs;
            }
            if (!string.IsNullOrWhiteSpace(incoming.Name) && incomingWins) existing.Name = incoming.Name.Trim();
            if (!string.IsNullOrWhiteSpace(incoming.Group) && incomingWins) existing.Group = incoming.Group.Trim();
            if (!string.IsNullOrWhiteSpace(incoming.SourceMachine) && (incomingWins || incomingCreatedWins)) existing.SourceMachine = incoming.SourceMachine.Trim();
            existing.Pinned = existing.Pinned || incoming.Pinned;
            existing.IsTemplate = existing.IsTemplate || incoming.IsTemplate;
            if (existing.ManualOrder <= 0 || (incoming.ManualOrder > 0 && incoming.ManualOrder < existing.ManualOrder)) existing.ManualOrder = incoming.ManualOrder;
        }

        private static void NormalizeMergedDatabase(ClipDatabase database)
        {
            if (database.Entries == null) database.Entries = new List<ClipEntry>();
            database.Version = Math.Max(1, database.Version);
            database.UpdatedUnixMs = TimeUtil.NowUnixMs();
            NormalizeDeletedEntries(database);
            ApplyDeletedEntries(database);

            var next = 1L;
            foreach (var entry in database.Entries
                .OrderBy(e => e.ManualOrder <= 0 ? long.MaxValue : e.ManualOrder)
                .ThenBy(e => e.CreatedUnixMs))
            {
                if (string.IsNullOrWhiteSpace(entry.Id)) entry.Id = Guid.NewGuid().ToString("N");
                if (entry.Name == null) entry.Name = string.Empty;
                if (entry.Group == null) entry.Group = string.Empty;
                if (entry.SourceMachine == null) entry.SourceMachine = string.Empty;
                if (entry.CreatedUnixMs == 0) entry.CreatedUnixMs = TimeUtil.NowUnixMs();
                if (entry.LastUsedUnixMs == 0) entry.LastUsedUnixMs = entry.CreatedUnixMs;
                entry.ManualOrder = next++;
            }
        }

        private static ClipEntry Clone(ClipEntry entry)
        {
            return new ClipEntry
            {
                Id = string.IsNullOrWhiteSpace(entry.Id) ? Guid.NewGuid().ToString("N") : entry.Id,
                Text = entry.Text ?? string.Empty,
                Name = entry.Name ?? string.Empty,
                Group = entry.Group ?? string.Empty,
                SourceMachine = entry.SourceMachine ?? string.Empty,
                CreatedUnixMs = entry.CreatedUnixMs,
                LastUsedUnixMs = entry.LastUsedUnixMs,
                Pinned = entry.Pinned,
                IsTemplate = entry.IsTemplate,
                ManualOrder = entry.ManualOrder
            };
        }

        private static void MergeDeletedEntries(ClipDatabase target, ClipDatabase source)
        {
            NormalizeDeletedEntries(target);
            if (source.DeletedEntries == null) return;
            foreach (var marker in source.DeletedEntries.Where(d => d != null && !string.IsNullOrWhiteSpace(d.Id)))
            {
                var existing = target.DeletedEntries.FirstOrDefault(d =>
                    string.Equals(d.Id, marker.Id, StringComparison.OrdinalIgnoreCase));
                if (existing == null)
                {
                    target.DeletedEntries.Add(Clone(marker));
                }
                else if (marker.DeletedUnixMs > existing.DeletedUnixMs)
                {
                    existing.DeletedUnixMs = marker.DeletedUnixMs;
                    existing.TextHash = marker.TextHash ?? string.Empty;
                    existing.SourceMachine = marker.SourceMachine ?? string.Empty;
                }
                else if (string.IsNullOrWhiteSpace(existing.TextHash) && !string.IsNullOrWhiteSpace(marker.TextHash))
                {
                    existing.TextHash = marker.TextHash;
                }
            }
            NormalizeDeletedEntries(target);
        }

        private static void NormalizeDeletedEntries(ClipDatabase database)
        {
            if (database.DeletedEntries == null)
            {
                database.DeletedEntries = new List<DeletedClipEntry>();
                return;
            }
            var now = TimeUtil.NowUnixMs();
            var cutoff = now - (long)TimeSpan.FromDays(90).TotalMilliseconds;
            foreach (var marker in database.DeletedEntries.Where(d => d != null && d.DeletedUnixMs <= 0))
            {
                marker.DeletedUnixMs = now;
            }
            database.DeletedEntries = database.DeletedEntries
                .Where(d => d != null && !string.IsNullOrWhiteSpace(d.Id) && d.DeletedUnixMs >= cutoff)
                .GroupBy(d => d.Id, StringComparer.OrdinalIgnoreCase)
                .Select(group => group.OrderByDescending(d => d.DeletedUnixMs).First())
                .ToList();
        }

        private static void ApplyDeletedEntries(ClipDatabase database)
        {
            if (database.Entries == null || database.DeletedEntries == null || database.DeletedEntries.Count == 0) return;
            database.Entries.RemoveAll(entry => IsDeleted(database, entry));
        }

        private static bool IsDeleted(ClipDatabase database, ClipEntry entry)
        {
            if (entry == null || database.DeletedEntries == null) return false;
            var entryChangedUnixMs = Math.Max(entry.CreatedUnixMs, entry.LastUsedUnixMs);
            var hash = ComputeTextHash(entry.Text);
            return database.DeletedEntries.Any(marker =>
                string.Equals(marker.Id, entry.Id, StringComparison.OrdinalIgnoreCase) ||
                (!string.IsNullOrWhiteSpace(marker.TextHash) &&
                 string.Equals(marker.TextHash, hash, StringComparison.OrdinalIgnoreCase) &&
                 (marker.DeletedUnixMs <= 0 || entryChangedUnixMs <= marker.DeletedUnixMs)));
        }

        private static string ComputeTextHash(string text)
        {
            using (var sha = SHA256.Create())
            {
                return string.Concat(sha.ComputeHash(Encoding.UTF8.GetBytes(text ?? string.Empty)).Select(b => b.ToString("x2")));
            }
        }

        private static DeletedClipEntry Clone(DeletedClipEntry marker)
        {
            return new DeletedClipEntry
            {
                Id = marker.Id ?? string.Empty,
                TextHash = marker.TextHash ?? string.Empty,
                DeletedUnixMs = marker.DeletedUnixMs,
                SourceMachine = marker.SourceMachine ?? string.Empty
            };
        }

        private static void DeleteFiles(IEnumerable<string> paths)
        {
            foreach (var path in paths)
            {
                try
                {
                    if (File.Exists(path))
                    {
                        File.Delete(path);
                    }
                }
                catch
                {
                }
            }
        }
    }
}
